src/lib/operators/table_scan.cpp

```
@@ -0,0 +1,206 @@
+    : AbstractOperator{in}, _column_id{column_id}, _scan_type{scan_type}, _search_value{search_value} {}
+
+ColumnID TableScan::column_id() const {
+  return _column_id;
+}
+
+ScanType TableScan::scan_type() const {
+  return _scan_type;
+}
+
+const AllTypeVariant& TableScan::search_value() const {
+  return _search_value;
+}
+
+std::shared_ptr<const Table> TableScan::_on_execute() {
+  const auto input_table = _left_input_table();
+
+  // TODO(Robert): What if last operator does not return anything?
+  Assert(input_table, "Performing a table scan without input does not work.");
+  const auto chunk_count = input_table->chunk_count();
+  auto const column_type = input_table->column_type(_column_id);
+  auto output_reference_segments = std::vector<std::shared_ptr<ReferenceSegment>>{};
+  output_reference_segments.reserve(chunk_count);
+
+  resolve_data_type(column_type, [&](auto type) {
+    using Type = typename decltype(type)::type;
+
+    for (auto chunk_id = ChunkID{0}; chunk_id < chunk_count; ++chunk_id) {
+      auto position_list = std::make_shared<PosList>();
+      const auto chunk = input_table->get_chunk(chunk_id);
+      // if chunk is empty, skip this chunk
+      if (!chunk->size()) {
+        continue;
+      }
+      const auto segment = chunk->get_segment(_column_id);
+
+      const auto value_segment = std::dynamic_pointer_cast<ValueSegment<Type>>(segment);
+      const auto dictionary_segment = std::dynamic_pointer_cast<DictionarySegment<Type>>(segment);
+      const auto reference_segment = std::dynamic_pointer_cast<ReferenceSegment>(segment);
+
+      if (value_segment) {
+        position_list = _tablescan_value_segment<Type>(value_segment, chunk_id);
+      } else if (dictionary_segment) {
+        position_list = _tablescan_dict_segment<Type>(dictionary_segment, chunk_id);
+      } else if (reference_segment) {
+        position_list = _tablescan_reference_segment<Type>(reference_segment, chunk_id);
+      }
+
```

Should we throw an error here, if none of the casts succeded?

---

yes